compiler/rustc_trait_selection/src/traits/const_evaluatable.rs

```diff
@@ -33,24 +31,25 @@ impl<'tcx> ConstUnifyCtxt<'tcx> {
     // Param(N) should unify with Param(T), substs: [Unevaluated("T2", [Unevaluated("T3", [Param(N)])])]
     #[inline]
     #[instrument(skip(self), level = "debug")]
-    fn try_replace_substs_in_root(
-        &self,
-        mut abstr_const: AbstractConst<'tcx>,
-    ) -> Option<AbstractConst<'tcx>> {
-        while let Node::Leaf(ct) = abstr_const.root(self.tcx) {
-            match AbstractConst::from_const(self.tcx, ct) {
-                Ok(Some(act)) => abstr_const = act,
+    fn try_replace_substs_in_root(&self, mut ct: ty::Const<'tcx>) -> Option<ty::Const<'tcx>> {
+        loop {
+            let uv = match ct.kind() {
+                ty::ConstKind::Unevaluated(uv) => uv,
+                ty::ConstKind::Error(_) => return None,
+                _ => break,
+            };
+            match self.tcx.bound_abstract_const(uv.shrink()) {
+                Ok(Some(sub_ct)) => ct = sub_ct,
                 Ok(None) => break,
                 Err(_) => return None,
             }
         }
-
-        Some(abstr_const)
+        Some(ct)
     }
 
     /// Tries to unify two abstract constants using structural equality.
     #[instrument(skip(self), level = "debug")]
-    pub fn try_unify(&self, a: AbstractConst<'tcx>, b: AbstractConst<'tcx>) -> bool {
+    pub fn try_unify(&self, a: ty::Const<'tcx>, b: ty::Const<'tcx>) -> bool {
```

Comment:
we want to remove `try_unify` but instead just use `infcx.at(cause, param_env).eq(ct_a, ct_b)` after expanding `ct_a` and `ct_b`

Reply:
this is still relevant